### PR TITLE
Add support for alternative container runtime in `kube-up.sh`

### PR DESCRIPTION
**Pull request by Random-Liu**
For https://github.com/kubernetes/features/issues/286.

This PR added 4 new environment variables in `kube-up.sh` to support alternative container runtime:
1) `KUBE_MASTER_EXTRA_METADATA` and `KUBE_NODE_EXTRA_METADATA`. Add extra metadata on master and node instance. With this we could specify different cloud-init for a different container runtime, and also add extra metadata for the new cloud-init, e.g. [master.yaml](https://github.com/Random-Liu/cri-containerd/blob/7d739662141cc137f8b1e82a9824b18be2e5df21/test/e2e/master.yaml)
2) `KUBE_CONTAINER_RUNTIME_ENDPOINT`. Specify different sock for different container runtime. It's only used when it's not empty.
3) `KUBE_LOAD_IMAGE_COMMAND`. Specify different load image command for different container runtime.

An example for cri-containerd:
```
export KUBE_MASTER_EXTRA_METADATA="user-data=${GOPATH}/src/github.com/kubernetes-incubator/cri-containerd/test/e2e/master.yaml,cri-containerd-configure-sh=${GOPATH}/src/github.com/kubernetes-incubator/cri-containerd/test/configure.sh"
export KUBE_NODE_EXTRA_METADATA="user-data=${GOPATH}/src/github.com/kubernetes-incubator/cri-containerd/test/e2e/node.yaml,cri-containerd-configure-sh=${GOPATH}/src/github.com/kubernetes-incubator/cri-containerd/test/configure.sh"
export KUBE_CONTAINER_RUNTIME="remote"
export KUBE_CONTAINER_RUNTIME_ENDPOINT="/var/run/cri-containerd.sock"
export KUBE_LOAD_IMAGE_COMMAND="/home/cri-containerd/usr/local/bin/cri-containerd load"
export NETWORK_POLICY_PROVIDER="calico"
```

Signed-off-by: Lantao Liu <lantaol@google.com>

```release-note
none
```
/cc @yujuhong @dchen1107 @feiskyer @mikebrow @abhi @mrunalp @runcom 
/cc @kubernetes/sig-node-pr-reviews

### CODE REVIEW

File: cluster/gce/config-default.sh [hunk 1]
```diff
@@ -81,8 +81,14 @@ NODE_IMAGE=${KUBE_GCE_NODE_IMAGE:-${GCI_VERSION}}
 NODE_IMAGE_PROJECT=${KUBE_GCE_NODE_PROJECT:-cos-cloud}
 NODE_SERVICE_ACCOUNT=${KUBE_GCE_NODE_SERVICE_ACCOUNT:-default}
 CONTAINER_RUNTIME=${KUBE_CONTAINER_RUNTIME:-docker}
+CONTAINER_RUNTIME_ENDPOINT=${KUBE_CONTAINER_RUNTIME_ENDPOINT:-}
+LOAD_IMAGE_COMMAND=${KUBE_LOAD_IMAGE_COMMAND:-docker load -i}
 RKT_VERSION=${KUBE_RKT_VERSION:-1.23.0}
 RKT_STAGE1_IMAGE=${KUBE_RKT_STAGE1_IMAGE:-coreos.com/rkt/stage1-coreos}
+# MASTER_EXTRA_METADATA is the extra instance metadata on master instance separated by colon.
```
Comment (yujuhong) [inline cluster/gce/config-default.sh]: nit: s/colon/colons

File: cluster/common.sh [hunk 2]
```diff
@@ -908,7 +911,6 @@ EOF
      [[ "${master}" == "false" && "${NODE_OS_DISTRIBUTION}" == "container-linux" ]]; then
     # Container-Linux-only env vars. TODO(yifan): Make them available on other distros.
     cat >>$file <<EOF
-KUBERNETES_CONTAINER_RUNTIME: $(yaml-quote ${CONTAINER_RUNTIME:-rkt})
```
Comment (yujuhong) [inline cluster/common.sh]: Should we keep this for backward compatibility for at least a release?

File: cluster/gce/config-default.sh [hunk 3]
```diff
@@ -81,8 +81,14 @@ NODE_IMAGE=${KUBE_GCE_NODE_IMAGE:-${GCI_VERSION}}
...
+MASTER_EXTRA_METADATA=${KUBE_MASTER_EXTRA_METADATA:-${KUBE_EXTRA_METADATA:-}}
+# MASTER_EXTRA_METADATA is the extra instance metadata on node instance separated by colon.
```
Comment (yujuhong) [inline cluster/gce/config-default.sh]: nit: s/colon/colons

File: cluster/gce/config-test.sh [hunk 4]
```diff
@@ -80,9 +80,15 @@ NODE_IMAGE=${KUBE_GCE_NODE_IMAGE:-${GCI_VERSION}}
 NODE_IMAGE_PROJECT=${KUBE_GCE_NODE_PROJECT:-cos-cloud}
 NODE_SERVICE_ACCOUNT=${KUBE_GCE_NODE_SERVICE_ACCOUNT:-default}
 CONTAINER_RUNTIME=${KUBE_CONTAINER_RUNTIME:-docker}
+CONTAINER_RUNTIME_ENDPOINT=${KUBE_CONTAINER_RUNTIME_ENDPOINT:-}
```
Comment (yujuhong) [inline cluster/gce/config-test.sh]: Just a question: when do we set KUBE_XXXXX variables?

**Comment by feiskyer** [comment]
Excellent work. 

One small question: is `KUBE_LOAD_IMAGE_COMMAND` required? Loading images is not part of CRI.

**Comment by Random-Liu** [comment]
> is KUBE_LOAD_IMAGE_COMMAND required? Loading images is not part of CRI.

Yeah, in `kube-up.sh` we need to load image for system containers. And different container runtimes need different commands.

File: cluster/gce/config-test.sh (hunk 4 above)
Comment (Random-Liu) [reply cluster/gce/config-test.sh]: Usually before we run `e2e-up.sh` and `kube-up.sh`.

File: cluster/gce/config-default.sh (hunk 3 above)
Comment (Random-Liu) [reply cluster/gce/config-default.sh]: Will do.
Actually should be commas. Sorry.

File: cluster/gce/config-default.sh (hunk 1 above)
Comment (Random-Liu) [reply cluster/gce/config-default.sh]: Will do.

File: cluster/common.sh (hunk 2 above)
Comment (Random-Liu) [reply cluster/common.sh]: OK. Will keep the container-linux side unchanged.

**Comment by Random-Liu** [comment]
@yujuhong Addressed comments.

**Comment by Random-Liu** [comment]
/test pull-kubernetes-bazel-test
--

**Comment by runcom** [comment]
Testing this out with cri-o, nice @Random-Liu

**Comment by Random-Liu** [comment]
> Testing this out with cri-o, nice @Random-Liu

You also need https://github.com/kubernetes/kubernetes/pull/54838

File: cluster/common.sh (hunk 2 above)
Comment (yujuhong) [reply cluster/common.sh]: I mean to accept both for now :-)

**Comment by Random-Liu** [comment]
Note that this PR only added alternative container runtime for COS and Ubuntu image.
If we want this on container-linux, we also need to change the script in container-linux. /cc @euank @yifan-gu

**Pull request merged**